### PR TITLE
[DC-733] Upgrade React to v18

### PR DESCRIPTION
**Pull request by nawatts**
- Upgrade React and React DOM to v18.
- Upgrade Testing Library packages to latest versions.
- Replace React Hooks Testing Library with `renderHook` from React Testing Library.
- Upgrade react-notifications-component to v4 (a version that supports React 18).
- [Replace `ReactDOM.render` with `ReactDOM.createRoot`](https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-client-rendering-apis).

### CODE REVIEW

File: src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.test.ts [hunk 1]
```diff
@@ -154,8 +154,7 @@ describe('AzureSubscriptionStep', () => {
 
   it('shows the spinner overlay while the call to list managed apps is in progress', async () => {
     renderAzureSubscriptionStep({});
-    const queryLoadingSpinner = () =>
-      screen.queryByRole((_, node: Element | null) => node?.getAttribute('data-icon') === 'loadingSpinner');
+    const queryLoadingSpinner = () => document.querySelector('[data-icon="loadingSpinner]');
```
Comment (nawatts) [inline src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.test.ts:157]: TS complained about this query. `queryByRole` takes a string parameter.

**Review by s-rubenstein — APPROVED** [review]
This looks very reasonable - no changes I didn't expect.

**Review by pshapiro4broad — APPROVED** [review]

**Comment by cahrens** [comment]
Do we want to give folks a heads-up and the chance to do some manual bashing?

**Review by cahrens — APPROVED** [review]
Clicked around in billing projects (including Highcharts usage) and Workspace dashboard. Code LGTM.

**Review by mlilynolting — APPROVED** [review]

**Pull request merged**